Sources/UserScript/UserScriptMessaging.swift

```diff
@@ -0,0 +1,263 @@
+        /// we **do not** forward any errors to the client
+        /// As far as the client is concerned, a `notification` is fire-and-forget
+        case .notify(let handler, let notification):
+            do {
+                try handler(notification.params, original) { _ in
+                    // handler ignored
+                }
+            } catch let error {
+                os_log("UserScriptMessaging: unhandled exception %s", type: .error, String(describing: error.localizedDescription))
+            }
+            return completion("{}")
+
+        /// Here the client will be expecting a response, so we always to produce one
+        /// We catch errors from handlers so that we can forward the response to clients with the correct context
+        ///
+        /// Most of the logic here is around ensuring we send either `result` or `error` in the response
+        /// Since that's how the Javascript side determines if the request was successful or not.
+        case .respond(let handler, let request):
+            do {
+                try handler(request.params, original) { encodableResponse in
+                    // handle an error
+                    guard let result = encodableResponse else {
+                        let resp = MessageErrorResponse.forRequest(request: request, error: "could not access encodable result").toJSON()
+                        return completion(resp)
+                    }
+
+                    // send the result
+                    if let resp = MessageResponse.toJSON(request: request, result: result) {
+                        return completion(resp)
+                    }
+
+                    let resp = MessageErrorResponse.forRequest(request: request, error: "could not convert result to json").toJSON()
+                    return completion(resp)
+                }
+            } catch let error {
+                let resp = MessageErrorResponse.forRequest(request: request, error: error.localizedDescription).toJSON()
+                return completion(resp)
+            }
+
+        /// We re-throw errors here to let consumers forward them as needed.
+        ///
+        /// Errors here are different to those caught in `.respond` above, because they are not
+        /// always tied to a response. They could be relating the incoming payload being incorrect etc.
+        case .error(let message):
+            let error = NSError(domain: "UserScriptMessaging", code: 0, userInfo: [NSLocalizedDescriptionKey: message])
+            throw error
+        }
+    }
+}
+
+/// An explicit format for specifying how a hostname should be matched
+public enum AllowedOrigin {
+    case etldPlus1(hostname: String)
+    case exact(hostname: String)
+}
+
+/// Force consumers to be explicit about the difference between accepting messages
+/// from 'all domains' vs 'only a subset of domains'
+public enum AllowedOrigins {
+    case all
+    case only(_: [AllowedOrigin])
```

Comment:
just by looking at this raw code I don't like the naming for those 2 enums:
1. enum should really be singular not plural as it doesn't aggregate stuff
2. because of similarity between these two
3. why is "only" an array?

how about we:
1. rename AllowedOrigin to HostnameMatchingRule 
2. rename "only" to only(rules: [HostnameMatchingRule])
3. rename AllowedOrigin to MessageOriginPolicy

these are just some ideas, I'm open for a discussion